src/heap.rs

```diff
@@ -20,71 +17,149 @@ impl HeapId {
+    }
+
+    fn write_u256(&mut self, start_address: u32, value: U256, pagepool: &mut PagePool) {
+        let (page_idx, offset_in_page) = address_to_page_offset(start_address);
+        let bytes_in_page = HEAP_PAGE_SIZE - offset_in_page;
+        let page = self.get_or_insert_page(page_idx, pagepool);
+
+        if bytes_in_page >= 32 {
+            value.to_big_endian(&mut page.0[offset_in_page..offset_in_page + 32]);
+        } else {
+            let mut bytes = [0; 32];
+            value.to_big_endian(&mut bytes);
+            let mut bytes_iter = bytes.into_iter();
 
-        value.to_big_endian(&mut self.0[start_address as usize..end]);
+            for (dst, src) in page.0[offset_in_page..].iter_mut().zip(bytes_iter.by_ref()) {
+                *dst = src;
+            }
+
+            let page = self.get_or_insert_page(page_idx + 1, pagepool);
+            for (dst, src) in page.0.iter_mut().zip(bytes_iter) {
+                *dst = src;
+            }
+        }
     }
 }
 
+#[inline(always)]
+fn address_to_page_offset(address: u32) -> (usize, usize) {
+    let offset = address as usize;
+    (offset >> 12, offset & (HEAP_PAGE_SIZE - 1))
+}
+
 impl HeapInterface for Heap {
     fn read_u256(&self, start_address: u32) -> U256 {
-        self.read_u256_partially(start_address..start_address + 32)
+        let (page_idx, offset_in_page) = address_to_page_offset(start_address);
+        let bytes_in_page = HEAP_PAGE_SIZE - offset_in_page;
+
+        if bytes_in_page >= 32 {
+            if let Some(page) = self.page(page_idx) {
+                U256::from_big_endian(&page.0[offset_in_page..offset_in_page + 32])
+            } else {
+                U256::zero()
+            }
+        } else {
+            let mut result = [0u8; 32];
+            if let Some(page) = self.page(page_idx) {
+                for (res, src) in result.iter_mut().zip(&page.0[offset_in_page..]) {
+                    *res = *src;
+                }
+            }
+            if let Some(page) = self.page(page_idx + 1) {
+                for (res, src) in result[bytes_in_page..].iter_mut().zip(&*page.0) {
+                    *res = *src;
+                }
+            }
+            U256::from_big_endian(&result)
+        }
     }
+
     fn read_u256_partially(&self, range: Range<u32>) -> U256 {
-        let end = (range.end as usize).min(self.0.len());
+        let (page_idx, offset_in_page) = address_to_page_offset(range.start);
+        let length = range.len();
+        let bytes_in_page = length.min(HEAP_PAGE_SIZE - offset_in_page);
 
-        let mut bytes = [0; 32];
-        for (i, j) in (range.start as usize..end).enumerate() {
-            bytes[i] = self.0[j];
+        let mut result = [0u8; 32];
+        if let Some(page) = self.page(page_idx) {
+            for (res, src) in result[..bytes_in_page]
+                .iter_mut()
+                .zip(&page.0[offset_in_page..])
+            {
+                *res = *src;
+            }
```

Comment:
Dumb question: Here and elsewhere, are bytewise assignment actually faster than `copy_from_slice`? This is surprising to me (intuitively, `copy_from_slice` would lead to a single non-overlapping memcpy).

Reply:
I haven't checked the assembly but I think this may help the compiler to directly write the data backwards instead of first copying it and then reversing. The loop should result in memcpy anyway if appropriate.